### PR TITLE
Handle SIGPIPE

### PR DESCRIPTION
**Pull request by wznmickey**
Because of https://stackoverflow.com/questions/65755853/simple-word-count-rust-program-outputs-valid-stdout-but-panicks-when-piped-to-he/65760807, our cli will not handle SIGPIPE and will panic.

Import https://lib.rs/crates/sigpipe to solve this problem.

Fix problems like https://github.com/typst/typst/issues/5439.

### CODE REVIEW

**Comment by PgBiel** [comment]
I believe replacing all instances of `println!` by (fallible) `writeln!`, as is done elsewhere, should be enough. Seems like there was an oversight with `typst fonts` in particular:
https://github.com/typst/typst/blob/5c37a1cfea2a08b08f4af34ba867748fd90c687e/crates/typst-cli/src/fonts.rs

**Comment by wznmickey** [comment]
Now changed.
Change both the println in fonts and query

File: crates/typst-cli/src/fonts.rs [hunk 1]
```diff
@@ -1,20 +1,23 @@
...
+        writeln!(out, "{name}:").unwrap();
         if command.variants {
             for info in infos {
                 let FontVariant { style, weight, stretch } = info.variant;
-                println!("- Style: {style:?}, Weight: {weight:?}, Stretch: {stretch:?}");
+                writeln!(
+                    out,
+                    "- Style: {style:?}, Weight: {weight:?}, Stretch: {stretch:?}"
+                )
+                .unwrap();
```
[the remark this answers is not in the document]
Comment (wznmickey) [reply crates/typst-cli/src/fonts.rs:20]: It is tested that now it works well.

**Comment by wznmickey** [comment]
There are two exist methods to handle the error.
First, use `unwrap()` like `writeln!(out, "{serialized}").unwrap();` in query.rs line 32.
Second, use `_` like `let _ = writeln!(out, " {}", self.0);` in download.rs line 28.

I think the `unwarp` one is better.

File: crates/typst-cli/src/fonts.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wznmickey) [reply crates/typst-cli/src/fonts.rs:20]: > I think unwrapping here might generate the same problem, but either way we should at least make `fonts` fallible and return an IO error instead of crashing.

https://github.com/typst/typst/pull/5444#issuecomment-2492409056

[the remark this answers is not in the document]
Comment (PgBiel) [reply crates/typst-cli/src/fonts.rs:20]: I apologize if that was confusing: I'm not proposing to ignore the error, I just think it would be more consistent with other commands if we made `fn fonts()` return a `Result` (say, with `io::Error` as the error type) and then used `writeln!(out, ...)?`.

[the remark this answers is not in the document]
Comment (wznmickey) [reply crates/typst-cli/src/fonts.rs:20]: Now, query() would return a Result, but it will just `unwarp` the error given in writeln. All other functions will give a hint on this error, but I am not sure what hint we could give if we have an error in write.

**Comment by wznmickey** [comment]
> Thanks. I noticed there are some other commands / functions using `eprintln!` as well, in particular `greet` and `update`, though I guess it doesn't matter as much for error printing (we can perhaps leave them to a separate discussion).

I committed https://github.com/wznmickey/typst/commit/9ce9c24fd2691ea5a0f50ac9f7c17c47cb80c804 as a possible approach in using `writeln`. It is not in this PR now.

File: crates/typst-cli/src/fonts.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (PgBiel) [reply crates/typst-cli/src/fonts.rs:20]: I'm thinking of completely removing unwrap from usages of `writeln!` and doing similarly to `print_diagnostics` as follows:

https://github.com/typst/typst/blob/5c37a1cfea2a08b08f4af34ba867748fd90c687e/crates/typst-cli/src/query.rs#L31-L33

A similar `.map_err()?` should be enough.

File: crates/typst-cli/src/fonts.rs [hunk 2]
```diff
@@ -1,21 +1,28 @@
-use typst::text::FontVariant;
-use typst_kit::fonts::Fonts;
-
 use crate::args::FontsCommand;
+use ecow::eco_format;
+use std::io::{self, Write};
+use typst::{diag::StrResult, text::FontVariant};
+use typst_kit::fonts::Fonts;
```
Comment (PgBiel) [inline crates/typst-cli/src/fonts.rs:5]: Please respect the import order (std -> external -> crate with a blank line between each group).

File: crates/typst-cli/src/fonts.rs [hunk 3]
```diff
@@ -1,21 +1,28 @@
...
 
 /// Execute a font listing command.
-pub fn fonts(command: &FontsCommand) {
+pub fn fonts(command: &FontsCommand) -> StrResult<()> {
     let fonts = Fonts::searcher()
         .include_system_fonts(!command.font_args.ignore_system_fonts)
         .search_with(&command.font_args.font_paths);
-
+    let mut out = io::stdout();
```
Comment (PgBiel) [inline crates/typst-cli/src/fonts.rs:12]: I think we can use `std::io::stdout()` instead of importing `io` just for this (seems to be how it's done in compile.rs).

File: crates/typst-cli/src/main.rs [hunk 4]
```diff
@@ -18,15 +18,14 @@ use std::io::{self, Write};
 use std::process::ExitCode;
 use std::sync::LazyLock;
 
+use crate::args::{CliArguments, Command};
+use crate::timings::Timer;
```
Comment (PgBiel) [inline crates/typst-cli/src/main.rs:22]: Please undo this change.

File: crates/typst-cli/src/query.rs [hunk 5]
```diff
@@ -12,7 +12,7 @@ use crate::args::{QueryCommand, SerializationFormat};
 use crate::compile::print_diagnostics;
 use crate::set_failed;
 use crate::world::SystemWorld;
-
+use std::io::{self, Write};
```
Comment (PgBiel) [inline crates/typst-cli/src/query.rs:15]: Please move this to the top, respecting the import order.

File: crates/typst-cli/src/query.rs [hunk 6]
```diff
@@ -28,7 +28,9 @@ pub fn query(command: &QueryCommand) -> HintedStrResult<()> {
         Ok(document) => {
             let data = retrieve(&world, command, &document)?;
             let serialized = format(data, command)?;
-            println!("{serialized}");
+            let mut out = io::stdout();
```
Comment (PgBiel) [inline crates/typst-cli/src/query.rs:31]: Same comment about `std::io::stdout()`

**Review by PgBiel — COMMENTED** [review]
Looking nice, thank you

**Comment by wznmickey** [comment]
All changed. Sorry, I did not notice the order of imports and just used the formatter directly before.

**Review by PgBiel — COMMENTED** [review]
I've tested the changes locally and noticed the error is printed as such now. This suggests that the unwrap was indeed going to cause the same problem as before in some cases:

```
./target/debug/typst fonts | head -n 5
Bitstream Vera Sans:
Bitstream Vera Sans Mono:
Bitstream Vera Serif:
Cascadia Code:
Cascadia Code NF:
error: failed to print fonts (Broken pipe (os error 32))
```

Overall, I'm not 100% sure on what would be the most appropriate option here. For one, it makes sense to have an error when something you're expecting to happen doesn't happen (printing fonts). OTOH, the broken pipe error is just way too easy to run into (even rather intentionally) for that to be of any good in this case.

Perhaps we could make a helper function which automatically drops an IO error if it's a SIGPIPE in this case? I know you suggested using a separate library, but I feel like it might be overkill, and the fact that it seems to have some sort of "global" effect makes me feel uneasy.

I was thinking of something like this (sketch - might not be the ideal solution):

```rs
pub fn suppress_sigpipe(error: std::io::Error) -> StrResult<()> {
  match error.kind() {
    std::io::ErrorKind::BrokenPipe => Ok(()),
    err => bail!("io error: {err}"),
  }
}

// usage:
writeln!(out, "Something")
   .or_else(suppress_sigpipe)?
```

Also, I do wonder if we should keep the current behavior of printing an error (but not crashing) for `query`, since (given there's only a single `writeln!`) it only errors if absolutely nothing prints, in which case I'd argue the query failed. Maybe we can change only fonts for now. (I'd also like to hear any other opinions.)

**Comment by wznmickey** [comment]
In your test cases, the `head -n 5` will stop after it outputs the first five lines and as a result the pipe is closed and we receive a SIGPIPE and in our code we will throw an error.
In our languages like C++, the program will immediately stop by default when it receives a SIGPIPE, while rust seems just return this error and we need to manually config it. I think maybe we could either
1. Follow what other programs do. Directly exit.
2. Throw an error.

**Comment by PgBiel** [comment]
Based on some further research and also taking a look at https://github.com/rust-lang/rust/issues/62569, I believe the best thing to do is to suppress SIGPIPE **for commands which we don't need to run to completion.** `typst fonts` is an example of that, as you may want to manipulate the fonts which were printed. But **it should not be suppressed for long-running commands** like `typst compile`. Hence why I think using the library with seemingly global behavior (the one in the linked stackoverflow post, IIUC) isn't ideal for us.

I guess for `typst compile` there's still the case where we're printing the final export to stdout though, when the output file is `-`. But in that case we can just suppress it at the export stage, if at all necessary.

Either way, if the worst that can happen is receive an error message that is easy to ignore (and isn't a panic), it sounds good to me. But for some (more informational) commands, like `typst fonts`, it seems just wrong, which would justify suppressing there.

`typst query` is in a bit of a gray area, since it does perform compilation, but its output is sort of designed to be processed. However, since it's all printed in one go, it only errors if _everything_ fails to print, so maybe we don't need to do much here.

Regardless, I've seen recommendations of always returning an exit code of 141 to indicate pipe failure... maybe we can use a dedicated error variant for this? Not sure.

Let's wait for some further discussion.

**Comment by wznmickey** [comment]
Since we have different functions for different command, is it possible to apply the global behavior function inside functions like `font`.
By the way, could we switch to stderr when stdout has some issues? stderr is also possibility be piped but this kind of cases could be rare.

**Comment by laurmaedje** [comment]
After reading https://stackoverflow.com/questions/65755853/simple-word-count-rust-program-outputs-valid-stdout-but-panicks-when-piped-to-he/65760807 and https://github.com/rust-lang/rust/issues/62569, it seems to me that using the `sigpipe` crate and not handling any errors would be the best way.

**Comment by wznmickey** [comment]
Revert to the first implementation.
Is it necessary to call this function in `font` and `query` function rather than the main function?

**Comment by laurmaedje** [comment]
Main function is fine. Thanks!
